[SwiftUI] Add a Feature Flags settings pane to SwiftBrowser

rr-codes opened the pull request:
#### 58fecd8fdf058dca6af95d7fc464cc8a399b8e3a
<pre>
[SwiftUI] Add a Feature Flags settings pane to SwiftBrowser
<a href="https://bugs.webkit.org/show_bug.cgi?id=299275">https://bugs.webkit.org/show_bug.cgi?id=299275</a>
<a href="https://rdar.apple.com/161076993">rdar://161076993</a>

Reviewed by Abrar Rahman Protyasha.

Like MiniBrowser, make it possible to change feature flags in SwiftBrowser, including a search field,
categorization, and a reset button!

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/_WebKit_SwiftUI/API/View+WebViewModifiers.swift:
(View.webViewScrollEdgeEffectStyle(_:for:)):
* Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
(WebPreferenceFeature.allowSmartLists): Deleted.
* Source/WebKit/_WebKit_SwiftUI/Implementation/EnvironmentValues+Extras.swift:
(EnvironmentValues.webViewScrollEdgeEffectStyleContext):
* Source/WebKit/_WebKit_SwiftUI/Implementation/ViewModifierContexts.swift:
(WebPreferenceContext.features): Deleted.
(WebPreferenceContext.isEmpty): Deleted.
* Source/WebKit/_WebKit_SwiftUI/Implementation/WebViewRepresentable.swift:

No need for ad-hoc view modifiers anymore.

(WebViewRepresentable.updatePlatformView(_:context:)):
* Tools/SwiftBrowser/Source/Extensions/SwiftUI+Extras.swift: Copied from Tools/SwiftBrowser/Source/Views/BrowserView.swift.

Add a way to project a non-optional Binding to a value for a Dictionary Binding.

* Tools/SwiftBrowser/Source/Extensions/_WKFeature+Extras.swift: Added.
(_WKFeature.id):
(WebFeatureStatus.allCases):
(WebFeatureStatus.description):
(WebFeatureCategory.allCases):
(WebFeatureCategory.description):
* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift:
(BrowserViewModel.updateWebPreferences):
* Tools/SwiftBrowser/Source/ViewModel/FeatureFlags.swift: Added.
(FeatureFlag.key):
(FeatureFlag.defaultValue):
(FeatureFlagsStorage.shared):
(FeatureFlagsStorage.resolve(_:)):
(FeatureFlagsStorage.update(mergingOldEntries:with:)):
* Tools/SwiftBrowser/Source/ViewModel/FeatureFlagsModel.swift: Copied from Source/WebKit/_WebKit_SwiftUI/Implementation/ViewModifierContexts.swift.
(FeatureFlagsModel.groups(filteredBy:)):
(FeatureFlagsModel.update(_:newValue:)):
* Tools/SwiftBrowser/Source/Views/BrowserView.swift:
(BrowserView.body):
* Tools/SwiftBrowser/Source/Views/ContentView.swift:
(ContentView.body):
* Tools/SwiftBrowser/Source/Views/SettingsView.swift:
(FeatureFlagToggle.value):
(FeatureFlagToggle.body):
(FeatureFlagsView.groupedFeatures):
(FeatureFlagsView.featureList):
(FeatureFlagsView.body):
(SettingsView.body):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Implement the feature flags settings pane, using UserDefaults as the source of truth.

Canonical link: <a href="https://commits.webkit.org/300410@main">https://commits.webkit.org/300410@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b3ec2b36ed1bb72177b6d8fdfa9d97a67bfdc44

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122323 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42027 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32709 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128908 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74422 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/79c2a3e7-4ff5-43dc-8c47-6c34f2596bf6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124199 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50621 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92977 "20 flakes 25 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61794 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/80d8e1eb-45ac-4b5a-9a16-f263a4bed32e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125275 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34089 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109525 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73636 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e817dfb2-49a7-4cd9-ac8b-52ba0f9df166) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33089 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27685 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72399 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/103675 "Build was cancelled. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (failure); re-run-api-tests (failure); Compiled WebKit (cancelled)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27892 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131650 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49263 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37486 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101530 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49637 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105745 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101401 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25748 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46768 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24896 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46025 "Build was cancelled. Recent messages:Printed configuration") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49120 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54862 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48589 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51939 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50270 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/SwiftBrowser/Source/ViewModel/FeatureFlags.swift:
```diff
@@ -0,0 +1,83 @@
+            defaults.set(newEntries[key], forKey: key)
+        }
+
+        // Flag was changed
```
nit: comments point out the obvious!

aprotyas commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 54;
```
This is a stray change right? Probably best to omit?

rr-codes committed:
[SwiftUI] Add a Feature Flags settings pane to SwiftBrowser
https://bugs.webkit.org/show_bug.cgi?id=299275
rdar://161076993

Reviewed by Abrar Rahman Protyasha.

Like MiniBrowser, make it possible to change feature flags in SwiftBrowser, including a search field,
categorization, and a reset button!

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/_WebKit_SwiftUI/API/View+WebViewModifiers.swift:
(View.webViewScrollEdgeEffectStyle(_:for:)):
* Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
(WebPreferenceFeature.allowSmartLists): Deleted.
* Source/WebKit/_WebKit_SwiftUI/Implementation/EnvironmentValues+Extras.swift:
(EnvironmentValues.webViewScrollEdgeEffectStyleContext):
* Source/WebKit/_WebKit_SwiftUI/Implementation/ViewModifierContexts.swift:
(WebPreferenceContext.features): Deleted.
(WebPreferenceContext.isEmpty): Deleted.
* Source/WebKit/_WebKit_SwiftUI/Implementation/WebViewRepresentable.swift:

No need for ad-hoc view modifiers anymore.

(WebViewRepresentable.updatePlatformView(_:context:)):
* Tools/SwiftBrowser/Source/Extensions/SwiftUI+Extras.swift: Copied from Tools/SwiftBrowser/Source/Views/BrowserView.swift.

Add a way to project a non-optional Binding to a value for a Dictionary Binding.

* Tools/SwiftBrowser/Source/Extensions/_WKFeature+Extras.swift: Added.
(_WKFeature.id):
(WebFeatureStatus.allCases):
(WebFeatureStatus.description):
(WebFeatureCategory.allCases):
(WebFeatureCategory.description):
* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift:
(BrowserViewModel.updateWebPreferences):
* Tools/SwiftBrowser/Source/ViewModel/FeatureFlags.swift: Added.
(FeatureFlag.key):
(FeatureFlag.defaultValue):
(FeatureFlagsStorage.shared):
(FeatureFlagsStorage.resolve(_:)):
(FeatureFlagsStorage.update(mergingOldEntries:with:)):
* Tools/SwiftBrowser/Source/ViewModel/FeatureFlagsModel.swift: Copied from Source/WebKit/_WebKit_SwiftUI/Implementation/ViewModifierContexts.swift.
(FeatureFlagsModel.groups(filteredBy:)):
(FeatureFlagsModel.update(_:newValue:)):
* Tools/SwiftBrowser/Source/Views/BrowserView.swift:
(BrowserView.body):
* Tools/SwiftBrowser/Source/Views/ContentView.swift:
(ContentView.body):
* Tools/SwiftBrowser/Source/Views/SettingsView.swift:
(FeatureFlagToggle.value):
(FeatureFlagToggle.body):
(FeatureFlagsView.groupedFeatures):
(FeatureFlagsView.featureList):
(FeatureFlagsView.body):
(SettingsView.body):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Implement the feature flags settings pane, using UserDefaults as the source of truth.

Canonical link: https://commits.webkit.org/300410@main
